fix kubectl cp command error

WanLinghao opened the pull request:
fix kubectl cp command error.

modified:   pkg/kubectl/cmd/cp.go

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
fix kubectl cp error.
it happens when copy directory to pod and the directory path ends with '/'.
for example:
kubectl cp /tmp/test/ test-pod:/tmp/test/
it will fail with 
error: stat /tmp/test/test: no such file or directory

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes 

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
```

xiangpengzhao commented on pkg/kubectl/cmd/cp.go:
```diff
@@ -195,6 +195,11 @@ func makeTar(filepath string, writer io.Writer) error {
+	len := len(filepath)
+	if filepath[len-1] == '/' {
+		filepath = filepath[0 : len-1]
+	}
```
Add a test case to prove this?

WanLinghao replied:
to prove it goes  correctly when the input ends up with '/' ?
by the way,thanks for the reply

xiangpengzhao replied:
yeah.

CaoShuFeng replied:
Yeap, a unit test for this would be better.

xiangpengzhao commented on pkg/kubectl/cmd/cp.go:
```diff
@@ -195,6 +195,11 @@ func makeTar(filepath string, writer io.Writer) error {
+
+	len := len(filepath)
+	if filepath[len-1] == '/' {
+		filepath = filepath[0 : len-1]
```
`0` can be omitted.

WanLinghao replied:
ok

CaoShuFeng replied:
Will this panic if len=0?

CaoShuFeng replied:
What about: strings.HasSuffix(filepath, "/")

xiangpengzhao commented:
/release-note-none
/ok-to-test

WanLinghao committed:
fix kubectl cp command error.
it happens when copy directory to pod and the directory
path ends with '/'.
for example:
kubectl cp /XX/XX/ XX-POD:/XX/XX

modified:  pkg/kubectl/cmd/cp.go
modified:  pkg/kubectl/cmd/cp_test.go

WanLinghao commented:
/test pull-kubernetes-e2e-kops-aws

CaoShuFeng commented:
/test pull-kubernetes-e2e-kops-aws

WanLinghao commented:
/test pull-kubernetes-e2e-kops-aws

WanLinghao commented:
@shiywang  @eparis PTAL

WanLinghao commented:
@brendanburns  please take a look thanks

CaoShuFeng commented:
What happens when user run command like this:
`kubectl cp / mypod:/tmp`

WanLinghao commented:
@CaoShuFeng  copy the whole filesystem on host machine to the pod

CaoShuFeng commented:
> @brendanburns please take a look thanks
@WanLinghao  you pingged the wrong people
s/brendanburns/brendandburns/

CaoShuFeng commented:
/assign @brendandburns

brendandburns commented on pkg/kubectl/cmd/cp.go:
```diff
@@ -195,6 +195,11 @@ func makeTar(filepath string, writer io.Writer) error {
 	tarWriter := tar.NewWriter(writer)
 	defer tarWriter.Close()
+
+	if strings.HasSuffix(filepath, "/") && filepath != "/" {
```
Rather than this, just do:

```go
filepath = path.Clean(filepath)
```

or just inline that:
```go
return recursiveTar(path.Dir(path.Clean(filepath)), ...)
```

CaoShuFeng replied:
Good.
path.Clean() will not truncate "/" into empty string.

WanLinghao replied:
good! thanks

brendandburns commented:
One comment to simplify the code.

CaoShuFeng commented:
/LGTM

CaoShuFeng commented:
/lgtm cancel
Please run hack/update-gofmt.sh

WanLinghao commented:
/test  pull-kubernetes-federation-e2e-gce

brendandburns commented:
/retest
/lgtm
/approve no-issue
